src/main/scala/se/nimsa/sbx/filtering/FilteringDAO.scala

```diff
@@ -0,0 +1,203 @@
+      getTagFilterByNameAction(tagFilter.name).flatMap {
+        _.map {t =>
+          replaceTagFilterTagPathAction(tagFilter.tags.map(tftp => TagFilterTagPath(-1, t.id, tftp))) andThen
+          updateTagFilterAction(tagFilterRow).map(_ => tagFilterRow.copy(id = t.id))
+        }.getOrElse {
+          insertAction
+        }
+      }
+    }
+    res.map(r => {
+      val res = tagFilter.copy(id = r.id)
+      res
+    })
+  }
+
+  def getAllSourceFilters: Future[Seq[SourceTagFilter]] = db.run(sourceFilterQuery.result)
+
+  def createOrUpdateSourceFilter(sourceFilter: SourceTagFilter): Future[SourceTagFilter] = {
+    db.run {
+      getSourceFilterAction(SourceRef(sourceFilter.sourceType, sourceFilter.sourceId)).flatMap {
+        _.map { sf =>
+          val updatedSourceFilter = sf.copy(tagFilterId = sourceFilter.tagFilterId)
+          updateSourceFilterAction(updatedSourceFilter).map(_ => updatedSourceFilter)
+        }.getOrElse {
+          insertSourceFilterAction(sourceFilter)
+        }
+      }
+    }
+  }
+
+  def getSourceFilter(sourceRef: SourceRef): Future[Option[SourceTagFilter]] = {
+    db.run(getSourceFilterAction(sourceRef))
+  }
+
+  def removeSourceFilter(id: Long): Future[Unit] =   db.run(sourceFilterQuery.filter(_.id === id).delete.map(_ => {}))
+
+  def removeSourceFilter(sourceRef: SourceRef): Future[Unit] =
+    db.run(
+      sourceFilterQuery
+        .filter(r => r.sourceType === sourceRef.sourceType.toString && r.sourceId === sourceRef.sourceId)
+        .delete.map(_ => {}))
+
+
+  def insertTagFilter(tagFilter: TagFilterSpec): Future[TagFilterSpec] = {
+    val tagFilterRow = TagFilter(tagFilter.id, tagFilter.name, tagFilter.tagFilterType)
+    val upsertAction = for {
+        tfr <- insertTagFilterAction(tagFilterRow)
+        _ <- replaceTagFilterTagPathAction(tagFilter.tags.map(tftp => TagFilterTagPath(-1, tfr.id, tftp)))
+      } yield tfr
+
+    db.run(upsertAction).map(r => tagFilter.copy(id = r.id))
+  }
+
+  def removeTagFilter(tagFilterId: Long): Future[Unit] = db.run(tagFilterQuery.filter(_.id === tagFilterId).delete.map(_ => {}))
+
+  def listTagFilters(startIndex: Long, count: Long): Future[Seq[TagFilter]] = db.run {
+    tagFilterQuery
+      .drop(startIndex)
+      .take(count)
+      .result
+  }
+
+  def listSourceTagFilters(startIndex: Long, count: Long): Future[Seq[SourceTagFilter]] = db.run {
+    sourceFilterQuery
+      .drop(startIndex)
+      .take(count)
+      .result
+  }
+
+  def getTagPathsByFilterId(filterId: Long): Future[Seq[TagFilterTagPath]] = db.run {
+    tagPathQuery.filter(_.tagFilterId === filterId).result
+  }
+
+  def getTagFilter(id: Long): Future[Option[TagFilterSpec]] = db.run {
+    val innerJoin = for {
+      tagFilters <- tagFilterQuery.filter(_.id === id)
+      tagPaths <- tagPathQuery if tagFilters.id === tagPaths.tagFilterId
+    } yield (tagFilters, tagPaths)
+    innerJoin.result
+  }.map(s => s.headOption.map(t => TagFilterSpec(t._1, s.map(_._2))))
```

Comment:
It wouldn't hurt to defined everything as actions with the goal of not `map`ping any Futures in this class, and running multiple actions in a transaction. Here's an example:
```
  def getTagFilter(id: Long): Future[Option[TagFilterSpec]] = {
    val innerJoin = for {
      tagFilters <- tagFilterQuery.filter(_.id === id)
      tagPaths <- tagPathQuery if tagFilters.id === tagPaths.tagFilterId
    } yield (tagFilters, tagPaths)
    val getAction = innerJoin.result.map(s => s.headOption.map(t => TagFilterSpec(t._1, s.map(_._2))))
    db.run(getAction.transactionally)
  }
```